### ISSUE TITLE
Error with MixSIAR and global environment

### ISSUE THREAD

**Issue by kelsey38** [brianstock/MixSIAR#135]
Hello, I am new to MixSIAR and have followed the instructions for putting the GUI on my computer through R Studio. Everything seems to run fine except for the fact that I am unable to call up the full distributions of any of my data (as in show(p.fac1[,1,1]) or median(p.fac1)).  I get an error that "object p.fac1 does not exist). I have seen a previous version of MixSIAR run on another computer and I noted that when you load the GUI, several more packages load in the earlier version than what happens on my computer, and they load a bunch of stuff into the global environment. Nothing shows up in the global environment on R Studio when I load the GUI. I also get the following error: "(rsession.exe:9268): Gtk-WARNING **: Could not find the icon 'gtk-file'. The 'hicolor' theme
was not found either, perhaps you need to install it.
You can get a copy from:
	http://icon-theme.freedesktop.org/releases"

What does any of this mean?

**Comment by brianstock** [comment]
Hi Kelsey,

If you're using the GUI, I believe `mixsiar` holds all the model objects (`"mixsiar" %in% ls()`), and you can try `mixsiar$jags.1$p.fac1` instead of `p.fac1` (https://github.com/brianstock/MixSIAR/issues/21). If `jags.1` exists (`"jags.1" %in% ls()`), then use `jags.1$BUGSoutput$sims.list$p.fac1`.

If you want to just type `p.fac1` you'll have to `attach.jags(jags.1)` or `attach.jags(mixsiar$jags.1)` first (https://github.com/brianstock/MixSIAR/issues/30, https://github.com/brianstock/MixSIAR/issues/80, https://github.com/brianstock/MixSIAR/issues/53).

Troubleshooting the GUI is hard and OS-dependent. I'd use the most recent version of MixSIAR. R Studio sometimes does not play nice with the GUI, so you can also try running `mixsiar_gui()` directly from R. If the GUI opens and the model runs, I'd call that a success and not worry about any GUI-related warnings like that!

**Comment by kelsey38** [comment]
Hello Brian,
Thanks for your help with this! I tried following the directions in the manual for deleting and reinstalling all the GTK+ related items, thinking that might be the issue. That did not help yesterday. Today I tried running the following after processing the output of the mixing models (you can see the code I ran plus the warnings/errors from the output):
> attach.jags(jags.1)
Error in attach.jags(jags.1) : could not find function "attach.jags"
In addition: Warning messages:
1: In cor(X, use = "pairwise.complete.obs") :
  the standard deviation is zero
2: In cor(X, use = "pairwise.complete.obs") :
  the standard deviation is zero
3: In cor(X, use = "pairwise.complete.obs") :
  the standard deviation is zero
4: In cor(X, use = "pairwise.complete.obs") :
  the standard deviation is zero
5: In cor(X, use = "pairwise.complete.obs") :
  the standard deviation is zero
6: In cor(X, use = "pairwise.complete.obs") :
  the standard deviation is zero
7: Removed 300 rows containing missing values (geom_bar). 
8: In cor(as.matrix(X[, -c(1, 2), drop = FALSE])) :
  the standard deviation is zero
9: Removed 2 rows containing missing values (geom_point). 
10: Removed 6 rows containing missing values (geom_point). 
> attach.jags(mixsiar$jags.1)
Error in attach.jags(mixsiar$jags.1) : 
  could not find function "attach.jags"
> attach(mixsiar$jags.1)
> p.fac1
Error: object 'p.fac1' not found
> mixsiar$jags.1$p.fac1
NULL
> ("mixsiar"%in%ls())
[1] FALSE
Does any of this make sense to you? Just so you know I have the most up-to-date versions of R, MixSIAR and R studio.

**Comment by brianstock** [comment]
[`attach.jags`](https://www.rdocumentation.org/packages/R2jags/versions/0.5-7/topics/attach.jags) is in the `R2jags` package. Should be loaded when you load MixSIAR via `library(MixSIAR)`.

Strange you're getting SD = 0 errors... Is this for your data or one of the examples? Let's make sure the examples work first:

```
library(MixSIAR)
mixsiar.dir <- find.package("MixSIAR")
source(paste0(mixsiar.dir,"/example_scripts/mixsiar_script_wolves.R"))
```

**Comment by kelsey38** [comment]
Hmm I think there's something wrong. I was running my data (getting the SD=0 errors) but when I did the wolves data, I didn't get the SD warnings, just the error that "object 'jags.1' is not found."  The R2jags package does not load when I load MixSIAR so that's an issue, and when I ran the script you posted above, R made the isospace plot of the wolves data and then gave this error:

Error in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y,  : 
  Metric information not available for this family/device

What would you suggest? And thank you very much for all your help!

**Comment by brianstock** [comment]
Do you have `R2jags` installed? Try `library(R2jags)`.

What is the output of `sessionInfo()`?

This has come up before, seems like it was either a problem with 1) old version of Mac OS, or 2) R Studio. Didn't have a problem using a newer Mac or running in the R console. See https://github.com/brianstock/MixSIAR/issues/78.

**Comment by kelsey38** [comment]
Here's library(R2jags) output:
> library(R2jags)
Loading required package: rjags
Loading required package: coda
Linked to JAGS 4.3.0
Loaded modules: basemod,bugs,dic,glm

Attaching package: ‘R2jags’

The following object is masked from ‘package:coda’:

    traceplot

> attach(jags.1)
Error in attach(jags.1) : object 'jags.1' not found

Here's sessionInfo output:
> sessionInfo()
R version 3.5.0 (2018-04-23)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] R2jags_0.5-7         rjags_4-6            coda_0.19-1          bindrcpp_0.2.2       gWidgetsRGtk2_0.0-86
[6] cairoDevice_2.24     gWidgets_0.0-54      RGtk2_2.20.34        MixSIAR_3.1.10      

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.16       bindr_0.1.1        pillar_1.2.2       compiler_3.5.0     RColorBrewer_1.1-2 plyr_1.8.4        
 [7] tools_3.5.0        boot_1.3-20        digest_0.6.15      tibble_1.4.2       gtable_0.2.0       lattice_0.20-35   
[13] pkgconfig_2.0.1    rlang_0.2.0        Matrix_1.2-14      GGally_1.3.2       yaml_2.1.18        parallel_3.5.0    
[19] SparseM_1.77       dplyr_0.7.4        stringr_1.3.0      ggmcmc_1.1         MatrixModels_0.4-1 tidyselect_0.2.4  
[25] grid_3.5.0         glue_1.2.0         reshape_0.8.7      R6_2.2.2           purrr_0.2.4        tidyr_0.8.0       
[31] magrittr_1.5       ggplot2_2.2.1      reshape2_1.4.3     scales_0.5.0       R2WinBUGS_2.1-21   mcmc_0.9-5        
[37] MASS_7.3-49        assertthat_0.2.0   abind_1.4-5        colorspace_1.3-2   labeling_0.3       quantreg_5.35     
[43] stringi_1.1.7      MCMCpack_1.4-2     lazyeval_0.2.1     munsell_0.4.3

My computer should be up to date (Running Windows 10) and I just re-installed R and R studio for this application (unless something got messed up with that). Should I uninstall R and R studio and try it all over again?

**Comment by brianstock** [comment]
R2jags function is `attach.jags`, not `attach`. 

I wouldn't reinstall R and R Studio... I would try running `library(MixSIAR); mixsiar_gui()` in the R console directly (not in R Studio).

And I would try the wolves example script version (see above). May also be your PDF device, can try saving png files instead of pdf.

**Comment by kelsey38** [comment]
Hi Brian,

Thanks again for your help with this. I tried running the wolves example in the console and found that once I used the code library(R2jags) after I processed the output I was finally able to access p.fac1 after using the attach.jags function. Not sure why this wasn't working in R studio but at least that problem is solved.

The other issue I'm having must have to do with my data then. I'm still getting the warning messages about sd's being 0 (see output at end of post). Do you know what these warnings mean? I've attached my source, consumer, and discrimination files so you can see how they're set up, and my models are contributions of various C sources to stream DIC (using Watershed as a fixed effect, with sources differing by watershed). I figured out that this error does not happen when I call "watershed" a random effect, but I'm only interested in a separate model for each watershed, not a global model.

[DIC_consumer.xlsx](https://github.com/brianstock/MixSIAR/files/1985228/DIC_consumer.xlsx)
[DIC_discrimination.xlsx](https://github.com/brianstock/MixSIAR/files/1985229/DIC_discrimination.xlsx)
[DIC_sources.xlsx](https://github.com/brianstock/MixSIAR/files/1985230/DIC_sources.xlsx)
 

Warning messages:
1: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
2: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
3: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
4: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
5: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
6: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
7: Removed 300 rows containing missing values (geom_bar).
8: In cor(as.matrix(X[, -c(1, 2), drop = FALSE])) :
  the standard deviation is zero
9: Removed 2 rows containing missing values (geom_point).
10: Removed 6 rows containing missing values (geom_point).
11: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
12: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
13: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
14: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
15: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
16: In cor(X, use = "pairwise.complete.obs") : the standard deviation is zero
17: Removed 300 rows containing missing values (geom_bar).
18: In cor(as.matrix(X[, -c(1, 2), drop = FALSE])) :
  the standard deviation is zero
19: Removed 2 rows containing missing values (geom_point).
20: Removed 6 rows containing missing values (geom_point).

**Comment by RoisinMcCallum** [comment]
Hi Kelsey, did you find a solution to the standard deviation= 0 issue? i am having the same problem with my data.

**Comment by kelsey38** [comment]
Hello, it's been a while since I did this, but I resolved the problem by one of these two methods (not sure which did it since I did both at the same time): First, I think I changed how the model was entered (it may have had to do with what I was listing as fixed/random effects, and I deleted one of the effects that I wasn't interested in that may have had sample sizes of n = 1 in some cases, leading to the error of SD = 0). Secondly, I had installed MixSIAR on a computer that was giving me issues because of the fact that I was using R studio and possibly the wrong version of R. When I moved to a computer that had MixSIAR installed by the University and also entered all of my model information in correctly, the error was resolved. Hope this is helpful!
